drivers/arm-cca/ukrsi/rsi.c

```diff
@@ -0,0 +1,288 @@
+#include <uk/arch/paging.h>
+#include <uk/plat/paging.h>
+#include <uk/plat/io.h>
+#include <uk/plat/common/sections.h>
+#include <uk/rsi.h>
+#include <kvm-arm64/image.h>
+
+unsigned long unprotected_mask;
+
+unsigned long uk_rsi_attestation_token_continue(unsigned long addr,
+						unsigned long *size)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_ATTESTATION_TOKEN_CONTINUE;
+	args.a1 = addr;
+
+	smccc_invoke(&args);
+	*size = args.a1;
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_attestation_token_init(unsigned long addr,
+					    unsigned long challenge[8])
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_ATTESTATION_TOKEN_INIT;
+	args.a1 = addr;
+	args.a2 = challenge[0];
+	args.a3 = challenge[1];
+	args.a4 = challenge[2];
+	args.a5 = challenge[3];
+	args.a6 = challenge[4];
+	args.a7 = challenge[5];
+	args.a8 = challenge[6];
+	args.a9 = challenge[7];
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_host_call(unsigned long addr)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_HOST_CALL;
+	args.a1 = addr;
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_ipa_state_get(unsigned long addr, unsigned char *ripas)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_IPA_STATE_GET;
+	args.a1 = addr;
+
+	smccc_invoke(&args);
+	*ripas = args.a1;
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_ipa_state_set(unsigned long base, unsigned long top,
+				   unsigned char ripas, unsigned long flags,
+				   unsigned long *new_base)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_IPA_STATE_SET;
+	args.a1 = base;
+	args.a2 = (top - base);
+	args.a3 = ripas;
+	args.a4 = flags;
+
+	smccc_invoke(&args);
+
+	*new_base = args.a1;
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_measurement_extend(unsigned long index, unsigned long size,
+					unsigned long value[8])
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_MEASUREMENT_EXTEND;
+	args.a1 = index;
+	args.a2 = size;
+	args.a3 = value[0];
+	args.a4 = value[1];
+	args.a5 = value[2];
+	args.a6 = value[3];
+	args.a7 = value[4];
+	args.a8 = value[5];
+	args.a9 = value[6];
+	args.a10 = value[7];
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_measurement_read(unsigned long index,
+				      unsigned long value[8])
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_MEASUREMENT_READ;
+	args.a1 = index;
+
+	smccc_invoke(&args);
+
+	value[0] = args.a1;
+	value[1] = args.a2;
+	value[2] = args.a3;
+	value[3] = args.a4;
+	value[4] = args.a5;
+	value[5] = args.a6;
+	value[6] = args.a7;
+	value[7] = args.a8;
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_realm_config(unsigned long config)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_REALM_CONFIG;
+	args.a1 = config;
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_version(void)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_VERSION;
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+void uk_rsi_init(struct ukplat_bootinfo *bi)
```

Comment:
Due to the nature of CCA it's important that `uk_rsi` functions return an error value, so that the handler can take a decision whether to proceed or not. In fact, I would even use `__check_result` here to force the caller to take action if `uk_rsi_init` fails. So this can be changed to:
```
int __check_result uk_rsi_init(struct ukplat_bootinfo *bi)
```
For the return values you can use somethning from `errno.h`, eg `-ENOTSUP`.